### PR TITLE
Validate entity id at server side

### PR DESCRIPTION
**Pull request by zengyan-amazon**
*Issue #, if available:*

*Description of changes:*
limit entity name allowed characters to numbers, letters, hyphen and underscore

test:
when creating internal user `a.b.c`
```
% curl -H "Content-Type: application/json" -H 'kbn-xsrf: true' -XPOST --include localhost:5601/nfh/api/v1/configuration/internalusers/a.b.c --cookie 'security_authentication=Fe26.2**d5ca53bbe4d1b6b6f67aadccd5754edc32003f069f16e88831add7e3ca88fb1c*99gIgfLsVDdzUs8cFpskZQ*9Ppr6Zy0JXSFz8z9_z0cGWwBRLXetInGl92Lbhc5mcNIGnccBCZ11KEbejDOrF9bmMWAgLg_VTnJQ2fpSV0uXXo62rp6LACUIcO2D6xuUshHhaD-5Sb6IwuynygzWBpsNbsQA_VLYq1UhIEzn2KQ9bvxCs5hJRHLZ5Vio3bzYj4DldBH-GLrSrSCqFAElvVBWbVb92TZyAdcGIzexvW22ksAbReZjHH3aJCAW-LLJhw**58a4f4b56956e60c9560c3f3b73b2e00feb1c39efcb165707cb09dbc5a1f6d61*a65_bMg-rjNWVCbKjZm5jExFLKtOwt2ImbfCxBbIAzc' -d'{ "password": "aaa"}'
HTTP/1.1 400 Bad Request
kbn-name: f8ffc2227bb6.ant.amazon.com
content-type: application/json; charset=utf-8
vary: origin
cache-control: private, no-cache, no-store, must-revalidate
set-cookie: security_authentication=Fe26.2**18844739027dfc367115a13b889531655d45411450555b9b26edf3ddc2439fc3*gYnSu_jpZodkY6lAOHqSaQ*KYdpb1f8P4dHQ7oygu5oFyzZM8n74s4uKoP_vHfcoZrenfHfthHsvy0CQPRaMU23KLdz3TVKT0GnZzqRzKn_NgxMULNPHh2dPCF1XpKO9_eZqXyJxvlR6y_2n5ZvQTEWk1hxeGBjdJLfvXhFXFP_XoTZQvRq6d2K3QbKFm8pRV9jT6wcdjgTed5pQ90JqAJWlogA0-9gwGSLWFLVNrAuWfkvDax_Xu23TVFwfipPRQA**e0e0340f8a146bd2a39f2f2ce00bacb95e24eb1816f8ae125c8877a2090474ff*XLgwPOd3nYBT2IKSlaWoNgPHjSGi7mlxTtBDjPoe-T4; HttpOnly; Path=/nfh
content-length: 100
date: Tue, 15 Sep 2020 06:02:53 GMT
Connection: keep-alive

{"statusCode":400,"error":"Bad Request","message":"[request params.id]: Invalid entity name or id."}
```

when updating tenant
```
% curl -H "Content-Type: application/json" -H 'kbn-xsrf: true' -XPOST --include localhost:5601/nfh/api/v1/multitenancy/tenant --cookie 'security_authentication=Fe26.2**d5ca53bbe4d1b6b6f67aadccd5754edc32003f069f16e88831add7e3ca88fb1c*99gIgfLsVDdzUs8cFpskZQ*9Ppr6Zy0JXSFz8z9_z0cGWwBRLXetInGl92Lbhc5mcNIGnccBCZ11KEbejDOrF9bmMWAgLg_VTnJQ2fpSV0uXXo62rp6LACUIcO2D6xuUshHhaD-5Sb6IwuynygzWBpsNbsQA_VLYq1UhIEzn2KQ9bvxCs5hJRHLZ5Vio3bzYj4DldBH-GLrSrSCqFAElvVBWbVb92TZyAdcGIzexvW22ksAbReZjHH3aJCAW-LLJhw**58a4f4b56956e60c9560c3f3b73b2e00feb1c39efcb165707cb09dbc5a1f6d61*a65_bMg-rjNWVCbKjZm5jExFLKtOwt2ImbfCxBbIAzc' -d'{ "username": "admin", "tenant": "<abc>"}'
HTTP/1.1 200 OK
kbn-name: f8ffc2227bb6.ant.amazon.com
content-type: text/html; charset=utf-8
vary: origin
cache-control: private, no-cache, no-store, must-revalidate
set-cookie: security_authentication=Fe26.2**2020f01b148f37f26eb267fdee80c92bc2e632c01fe9eea0c570486d92172bd8*-g0WD2YOuuHinD8As5WkyQ*VEB793YXymyAhqiJ2BK1L0r1vLN2Yf_gVkrq_PvMGxFYG0B1fhpuHEpEoMlTlbL6Jb65niwY8ftWZbwjFvQ56QYlr6YanIsg_x4I94ev1Bq56JFwQjtJyoZRvFIC9fg55T5s89Nnjk4DH_20zJfc4trQ9jL1DyFofYQT2Tw73Vd7jHqKTZ3NN362LwwhIvImhjFBUBtqAMilyIBoy4itCAy-0moWJ3PCtms3q9bOSGc**45b9c1a3dd4cd1547b7aa67f195101a68bb8aac6aceebf861a4c3e29e9fe2152*1477Ig-CvaVeAcL1G19UTBHo_FSiApjOxSORZjCl9Bw; HttpOnly; Path=/nfh
content-length: 11
date: Tue, 15 Sep 2020 06:00:50 GMT
Connection: keep-alive

&lt;abc&gt;
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/routes/index.ts [hunk 1]
```diff
@@ -82,6 +84,12 @@ export function defineRoutes(router: IRouter) {
     inputSchema.validate(requestBody); // throws error if validation fail
   }
 
+  function validateEntityId(resourceName: string) {
+    if (!RESOURCE_ID_REGEX.test(resourceName)) {
+      return 'Invalid entity name or id.';
```
Comment (yuxitang-amzn) [inline server/routes/index.ts:89]: 1. Do we want to handle the empty?
2. Do we want to tell user the valid char set?

Comment (zengyan-amazon) [reply server/routes/index.ts:89]: we don't allow empty the regular express requires some letter to match

And we have hint on the UI to tell users what characters are allowed, so I don't plan to to implement logic to identity invalid characters

Comment (yuxitang-amzn) [reply server/routes/index.ts:89]: Agreed.

**Review by yuxitang-amzn — APPROVED** [review]

File: server/routes/index.ts (hunk 1 above)
Comment (yuxitang-amzn) [reply server/routes/index.ts:89]: Agreed.

File: server/multitenancy/routes.ts [hunk 2]
```diff
@@ -52,7 +55,7 @@ export function setupMultitenantRoutes(
       cookie.tenant = tenant;
```
Comment (tianleh) [inline server/multitenancy/routes.ts:55]: Shall we encode in line 45 so that 55 and 58 will both get the encoded value?

Comment (zengyan-amazon) [reply server/multitenancy/routes.ts:55]: We can do that, while the most concern is to not producing HTML/script when responding to browser side, so encode at the last step is safer to me.

File: server/routes/index.ts (hunk 1 above)
Comment (zengyan-amazon) [reply server/routes/index.ts:86]: updated the regular expression to include unicode characters

Comment (yuxitang-amzn) [reply server/routes/index.ts:86]: Curious about the reason not using builtin regex like `/\p{L}+/u`? What is the difference?
And could we apply the length limit 2~50 here? BTW a unicode counts as 1 char or 2 chars?

**Pull request merged**